internal/serialization/serialization.go

```diff
@@ -51,26 +51,36 @@ func NewService(serializationConfig *pubserialization.Config) (*Service, error)
 // ToData serializes an object to a Data.
 // It can safely be called with a Data. In that case, that instance is returned.
 // If it is called with nil, nil is returned.
-func (s *Service) ToData(object interface{}) (pubserialization.Data, error) {
-	if _, ok := object.(*SerializationData); ok {
-		return object.(*SerializationData), nil
+func (s *Service) ToData(object interface{}) (r *Data, err error) {
+	defer func() {
+		if rec := recover(); rec != nil {
+			err = hzerror.MakeError(rec)
```

Comment:
Panic is considered a bad practice when it comes to normal error handling. Panic is meant to be used for non-recoverable error cases. What's the benefit in using it here? The same question applies to some other places in this file, like `ToObject()` method.

Reply:
If it's not possible to recover from an error (e.g., the type of some field is int but the value is not) what's the point of returning an error? There's nothing the API user can do about it. Also, that kind of errors are exceptions. It's quite cumbersome to return errors from each `WriteXXX` or `ReadXXX` function, and it is probably less performant (need to benchmark).

We recover from panics and return an error to the user, so I think as long as the panics are internal in our API, there's no problem using them.

Reply:
> If it's not possible to recover from an error (e.g., the type of some field is int but the value is not) what's the point of returning an error?

This code recovers from an error by handling the panic and returning it to the end user. So, I don't think I understand this point.

Also, serializers may be based on user code and a user may handle the panic in their own for their own purpose. In that case, user handler will "swallow" the unchecked error (let's call it this way), the recovery mechanism in this module may not work as expected.

As for performance considerations, I don't think error handling impacts performance, while frequent panic handling may certainly impact it (not that it's the main reason not to use it, although).

Reply:
This is a good article about about panics: https://eli.thegreenplace.net/2018/on-the-uses-and-misuses-of-panics-in-go/ Quoting: 

> It's a good practice for packages not to panic in their public interfaces. Rather, the public-facing functions and methods should recover from panics internally and translate them to error messages. This makes panics friendlier to use" 

That's exactly what I am doing here. Note that the author gives an example from Go standard lib, which doesn't check for errors at each step but relies on panics in the "The siren call of panic" section. He also criticizes that piece of code, since it tries to "discriminate between different kids of panics". I'm not doing that, the error from the panic is returned verbatim.

> This code recovers from an error by handling the panic and returning it to the end user. So, I don't think I understand this point.

What I mean is, if a problem occurs during serialization/deserialization, it doesn't make sense to try to return to a state where it is possible to continue serialization/deserialization. Assume the serializer expects the following
```
bar struct {
	foo struct {
		x int
	}
  	b bool
}
```
But it receives:
```
bar struct {
	foo struct {
              x bool
	}
}
```
What can the deserializer do apart from returning something like: "expected int but received bool" or even "invalid bar"?

> Also, serializers may be based on user code and a user may handle the panic in their own for their own purpose. In that case, user handler will "swallow" the unchecked error (let's call it this way), the recovery mechanism in this module may not work as expected.

I don't see why the user would catch a panic in their serializer. But if they do, that would be their problem.

> As for performance considerations, I don't think error handling impacts performance, while frequent panic handling may certainly impact it

Error handling may impact the performance, since every step should be checked for errors and "bubble up" the error several layers in case there is an error or not. There can be an arbitrary number of error checks during serialization/deserialization of a single value. Yes, panic handling should not be frequent, they should really be "exceptional" like it is in this PR: once in serialization and once in deserialization.

Reply:
> What can the deserializer do apart from returning something like: "expected int but received bool" or even "invalid bar"?

Obviously, it can return an error to the caller.

> I don't see why the user would catch a panic in their serializer.

For pretty much the same reason as you're doing it.

As for performance considerations, not taken branches in hot loops are cheap.

The error handling was already there, so it's not that this change cuts any corners. That's why I asked the question. Anyway, please consider this comment to be a nit.

Reply:
Actually, the previous situation was a bit different. Some of the serialization functions didn't return `error`s, some of did. The errors were ignored in some cases, like:
```
func (o *ObjectDataOutput) WriteByte(v byte) error {
	o.EnsureAvailable(bufutil.ByteSizeInBytes)
	WriteUInt8(o.buffer, o.position, v)
	o.position += bufutil.ByteSizeInBytes
	return nil
}

func (o *ObjectDataOutput) WriteBool(v bool) {
	o.EnsureAvailable(bufutil.BoolSizeInBytes)
	WriteBool(o.buffer, o.position, v)
	o.position += bufutil.BoolSizeInBytes
}
```